internal/users/db/migration.go

```diff
@@ -0,0 +1,252 @@
+
+		log.Debugf(context.Background(), "Migrating group %v", group.Name)
+		if err := insertGroup(tx, group); err != nil {
+			return err
+		}
+	}
+
+	// Add users to groups
+	for _, u := range bboltUsers {
+		groups, err := bboltDB.UserGroups(u.UID)
+		if errors.Is(err, bbolt.NoDataFoundError{}) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		for _, g := range groups {
+			if err := addUserToGroup(tx, u.UID, g.GID); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Add user to local groups
+	for _, u := range bboltUsers {
+		localGroups, err := bboltDB.UserLocalGroups(u.UID)
+		if errors.Is(err, bbolt.NoDataFoundError{}) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		for _, g := range localGroups {
+			if err := addUserToLocalGroup(tx, u.UID, g); err != nil {
+				return err
+			}
+		}
+	}
+
+	log.Debug(context.Background(), "Done migrating data from bbolt to SQLite")
+	return nil
+}
+
+type schemaMigration struct {
+	description string
+	migrate     func(*Manager) error
+}
+
+var schemaMigrations = []schemaMigration{
+	{
+		description: "Migrate to lowercase user and group names",
+		migrate: func(m *Manager) error {
+			users, err := m.AllUsers()
+			if err != nil {
+				return fmt.Errorf("failed to get users from database: %w", err)
+			}
+
+			var oldNames, newNames []string
+			for _, u := range users {
+				oldNames = append(oldNames, u.Name)
+				newNames = append(newNames, strings.ToLower(u.Name))
+			}
+
+			if err := renameUsersInUserGroupFile(oldNames, newNames); err != nil {
+				return fmt.Errorf("failed to rename users in %s file: %w", userGroupFile, err)
+			}
+
+			query := `UPDATE users SET name = LOWER(name);
+					  UPDATE groups SET name = LOWER(name);`
+			_, err = m.db.Exec(query)
+			return err
+		},
+	},
+}
+
+func (m *Manager) maybeApplyMigrations() error {
+	currentVersion, err := getSchemaVersion(m.db)
+	if err != nil {
+		return err
+	}
+
+	if currentVersion >= len(schemaMigrations) {
+		return nil
+	}
+
+	log.Debugf(context.Background(), "Schema version before migrations: %d", currentVersion)
+
+	v := 0
+	for _, migration := range schemaMigrations {
+		v++
+		if currentVersion >= v {
+			continue
+		}
+
+		log.Infof(context.Background(), "Applying schema migration: %s", migration.description)
+		if err := migration.migrate(m); err != nil {
+			return fmt.Errorf("error applying schema migration: %w", err)
+		}
+
+		if err := setSchemaVersion(m.db, v); err != nil {
+			return fmt.Errorf("failed to update schema version: %w", err)
+		}
+	}
+
+	log.Debugf(context.Background(), "Schema version after migrations: %d", v)
+
+	return nil
+}
+
+// renameUsersInUserGroupFile renames users in the /etc/group file.
+func renameUsersInUserGroupFile(oldNames, newNames []string) error {
+	// Note that we can't use gpasswd here because it checks for the existence of the user, which causes an NSS request
+	// being sent to authd, but authd is not ready yet because we are still migrating the database.
```

Comment:
I'm still wondering if  we could try to use it: we are in control of our NSS service, so basically we can just not start it until the migration is ready, and this would prevent the self-call... no?